pkg/api/datasources.go

```diff
@@ -706,7 +732,47 @@ func (hs *HTTPServer) CallDatasourceResourceWithUID(c *contextmodel.ReqContext)
 		return
 	}
 
-	hs.callPluginResourceWithDataSource(c, plugin.ID, ds)
+	// Check the cache first. If it's a hit, try to write it to the response
+	resp := hs.queryCachingService.HandleResourceRequest(c.Req.Context(), c.Req)
+	if isCacheHit(resp) {
+		if err := writeResourceResponse(c.Resp, resp); err != nil {
+			hs.log.Error("error trying to return cached resource response: %s", err.Error())
+		} else {
+			return
+		}
+	}
+	resp.WriteHeadersToResponse(&c.Resp)
+	// No cache hit, process resource request as usual
+	hs.callPluginResourceWithDataSource(c, plugin.ID, ds, resp.UpdateCacheFn)
+}
+
+func writeResourceResponse(w http.ResponseWriter, response caching.CachedResourceDataResponse) error {
```

Comment:
This function was copied from the old enterprise impl for writing the response in the middleware